### PR TITLE
qt6: qtdeclarative: add dependency to qtshadertools & other fixes

### PR DESCRIPTION
**Pull request by tenzap**
* qtdeclarative: add dependency to qtshadertools
    - Before that change, configure reported:
      `Note: Qt Quick modules not built due to not finding the qtshadertools 'qsb' tool.`
    - That led to the non-availability of Qt::QuickControls2 and QtQuick modules for other projects
* Increment revision on dependants of qtdeclarative
* Fix `Unknown CMake command "qt_internal_add_qml_module"` issue
    - by removing the reinplace of ${CMAKE_CURRENT_LIST_DIR}
      which was responsible for the non installation of some cmake files
      See also: https://bugreports.qt.io/browse/QTBUG-109003
* Increment revision of qtbase and remaining modules
* Remove patches that are not referenced anymore
    - patch-qiosurfacegraphicsbuffer.h.diff
    - patch-qmake_flags.diff

Original configure log:
```
Executing:  cd "/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_aqua_qt6/qt6-qtdeclarative/work/build" && /opt/local/libexec/qt6/bin/qt-configure-module /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_aqua_qt6/qt6-qtdeclarative/work/qtdeclarative-everywhere-src-6.2.4 -- -DCMAKE_C_COMPILER=/opt/local/bin/clang-mp-15 -DCMAKE_CXX_COMPILER=/opt/local/bin/clang++-mp-15 -DCMAKE_OBJC_COMPILER=/opt/local/bin/clang-mp-15 -DCMAKE_OBJCXX_COMPILER=/opt/local/bin/clang++-mp-15 
'/opt/local/libexec/qt6/bin/qt-cmake-private' '-DCMAKE_C_COMPILER=/opt/local/bin/clang-mp-15' '-DCMAKE_CXX_COMPILER=/opt/local/bin/clang++-mp-15' '-DCMAKE_OBJC_COMPILER=/opt/local/bin/clang-mp-15' '-DCMAKE_OBJCXX_COMPILER=/opt/local/bin/clang++-mp-15' '/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_aqua_qt6/qt6-qtdeclarative/work/qtdeclarative-everywhere-src-6.2.4'
-- The CXX compiler identification is Clang 15.0.5
-- The C compiler identification is Clang 15.0.5
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/local/bin/clang++-mp-15 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/local/bin/clang-mp-15 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Performing Test HAVE_STDATOMIC
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE  
-- Found OpenGL: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenGL.framework   
-- Could NOT find Qt6ShaderTools (missing: Qt6ShaderTools_DIR)
CMake Warning at /opt/local/libexec/qt6/lib/cmake/Qt6/Qt6Config.cmake:259 (message):
  Failed to find Qt component "ShaderTools".

  Expected Config file at
  "/opt/local/libexec/qt6/lib/cmake/Qt6ShaderTools/Qt6ShaderToolsConfig.cmake"
  does NOT exist

Call Stack (most recent call first):
  CMakeLists.txt:45 (find_package)


-- CMAKE_BUILD_TYPE was set to: 'Release'
-- Check for feature set changes
-- The OBJC compiler identification is Clang 15.0.5
-- Detecting OBJC compiler ABI info
-- Detecting OBJC compiler ABI info - done
-- Check for working OBJC compiler: /opt/local/bin/clang-mp-15 - skipped
-- The OBJCXX compiler identification is Clang 15.0.5
-- Detecting OBJCXX compiler ABI info
-- Detecting OBJCXX compiler ABI info - done
-- Check for working OBJCXX compiler: /opt/local/bin/clang++-mp-15 - skipped
-- Tool 'Qt6::qmltyperegistrar' will be built from source.
-- Could NOT find LTTngUST (missing: LTTNGUST_LIBRARIES LTTNGUST_INCLUDE_DIRS) 
-- Found Python: /opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10 (found version "3.10.8") found components: Interpreter 
-- Performing Test HAVE_pointer_32bit
-- Performing Test HAVE_pointer_32bit - Failed
-- Performing Test HAVE_pointer_64bit
-- Performing Test HAVE_pointer_64bit - Success
-- Performing Test HAVE_arm_thumb
-- Performing Test HAVE_arm_thumb - Failed
-- Performing Test HAVE_arm_fp
-- Performing Test HAVE_arm_fp - Success
-- Using installed syncqt found at: /opt/local/libexec/qt6/bin/syncqt.pl
-- Running syncqt for module: 'QtQml' 
-- Could NOT find LTTngUST (missing: LTTNGUST_LIBRARIES LTTNGUST_INCLUDE_DIRS) 
-- Performing Test HAVE_DASH_UNDEFINED_SYMBOLS
-- Performing Test HAVE_DASH_UNDEFINED_SYMBOLS - Success
-- Performing Test HAVE_DASH_DASH_NO_UNDEFINED
-- Performing Test HAVE_DASH_DASH_NO_UNDEFINED - Failed
-- Running syncqt for module: 'QtQmlModels' 
-- Running syncqt for module: 'QtQmlCore' 
-- Running syncqt for module: 'QtQmlWorkerScript' 
-- Running syncqt for module: 'QtQmlLocalStorage' 
-- Running syncqt for module: 'QtQmlXmlListModel' 
-- Running syncqt for module: 'QtLabsSettings' 
-- Running syncqt for module: 'QtLabsQmlModels' 
-- Running syncqt for module: 'QtLabsFolderListModel' 
-- Running syncqt for module: 'QtPacketProtocol' 
-- Running syncqt for module: 'QtQmlDom' 
-- Running syncqt for module: 'QtQmlCompiler' 
-- Tool 'Qt6::qmlcachegen' will be built from source.
-- Running syncqt for module: 'QtQmlDebug' 
-- Tool 'Qt6::qmldom' will be built from source.
-- Tool 'Qt6::qmllint' will be built from source.
-- Tool 'Qt6::qmlimportscanner' will be built from source.
-- Tool 'Qt6::qmlformat' will be built from source.
-- Tool 'Qt6::qmlprofiler' will be built from source.
-- The following packages have been found:

 * PkgConfig
 * Qt6BuildInternals (required version >= 6.2.4)
 * Qt6CoreTools (required version >= 6.2.4)
 * Qt6Core (required version >= 6.2.4)
 * OpenGL
 * Qt6GuiTools (required version >= 6.2.4)
 * Qt6DBusTools (required version >= 6.2.4)
 * Qt6Gui (required version >= 6.2.4)
 * Qt6Network (required version >= 6.2.4)
 * Qt6WidgetsTools (required version >= 6.2.4)
 * Qt6Widgets (required version >= 6.2.4)
 * Qt6OpenGL (required version >= 6.2.4)
 * Qt6OpenGLWidgets (required version >= 6.2.4)
 * Qt6Sql (required version >= 6.2.4)
 * Qt6Concurrent (required version >= 6.2.4)
 * Qt6Test (required version >= 6.2.4)
 * Qt6 (required version >= 6.2.4)
 * Python

-- The following OPTIONAL packages have not been found:

 * Qt6ShaderToolsTools (required version >= 6.2.4)
 * Qt6ShaderTools (required version >= 6.2.4)
 * LTTngUST

Configure summary:

Qt QML:
  QML network support .................... yes
  QML debugging and profiling support .... yes
  QML just-in-time compiler .............. yes
  QML sequence object .................... yes
  QML XML http request ................... yes
  QML Locale ............................. yes
Qt QML Models:
  QML list model ......................... yes
  QML delegate model ..................... yes
Qt Quick:
  Qt Quick support ....................... no

Note: Qt Quick modules not built due to not finding the qtshadertools 'qsb' tool.

-- Configuring done
```

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS x.y
Xcode x.y / Command Line Tools x.y.z

Check through CI build.

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [ ] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by tenzap** [comment]
QuickControls2 gets installed by qtdeclarative but now, qtwebchannel is failing with:

```
2022-11-28T09:29:43.0032400Z CMake Error at src/multimediaquick/CMakeLists.txt:9 (qt_internal_add_qml_module):
2022-11-28T09:29:43.0032860Z   Unknown CMake command "qt_internal_add_qml_module".
2022-11-28T09:29:43.0033030Z 
2022-11-28T09:29:43.0033040Z 
2022-11-28T09:29:43.0037020Z -- Configuring incomplete, errors occurred!
```

qt_internal_add_qml_module is in `<installdir>/lib/cmake/Qt6Qml/Qt6QmlBuildInternals.cmake`but for some reason this file is not installed on macports (I checked and in Debian 6.2.4 it is installed while on macports it is not).

I added a possible patch that would force installing Qt6QmlBuildInternals.cmake. CI outcome will tell if this works.

**Comment by tenzap** [comment]
I created a bug report at Qt: https://bugreports.qt.io/browse/QTBUG-109003

**Comment by tenzap** [comment]
> I added a possible patch that would force installing Qt6QmlBuildInternals.cmake. CI outcome will tell if this works.

Another problem appears: 
```
2022-11-28T13:08:47.2342060Z CMake Error at /opt/local/libexec/qt6/lib/cmake/Qt6Qml/Qt6QmlBuildInternals.cmake:297 (qt6_add_qml_module):
2022-11-28T13:08:47.2343020Z   Unknown CMake command "qt6_add_qml_module".
2022-11-28T13:08:47.2343320Z Call Stack (most recent call first):
2022-11-28T13:08:47.2344000Z   src/imports/qmlwebsockets/CMakeLists.txt:1 (qt_internal_add_qml_module)
2022-11-28T13:08:47.2344200Z 
```
File `Qt6QmlMacros.cmake` is then missing too.

**Comment by reneeotten** [comment]
Yes, that's the same problem we have always been having... just increasing the revision of all subports isn't going to do anything.

**Comment by tenzap** [comment]
I increased the revision of all modules that depend on qtdeclarative since qtdeclarative will also build all qtquick with all these changes. Qtquick availability may change the output of the packages directing on it, no?

**Comment by tenzap** [comment]
@reneeotten,according to the [comment in QTBUG-109003](https://bugreports.qt.io/browse/QTBUG-109003?focusedCommentId=692790&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-692790), concerning the `qt_internal_add_qml_module` issue, maybe the problem is created by this line
https://github.com/macports/macports-ports/blob/master/aqua/qt6/Portfile#L1324
`reinplace -q "s|\\\${CMAKE_CURRENT_LIST_DIR}|${qt_cmake_module_dir}/${cmakedir}|g" ${cmakefile}`

The macports' tbz2 package of qtbase shows that the original content is replaced by:
```
    if (EXISTS "/opt/local/libexec/qt6/lib/cmake/Qt6/${INSTALL_CMAKE_NAMESPACE}${target}Macros.cmake")
```
Thus qt won't install `Qt6QmlMacros.cmake` & `Qt6QmlBuildInternals.cmake`

Is this code snippet actually needed in the Portfile? Why? @jhoyt4 maybe?

**Comment by jhoyt4** [comment]
> @reneeotten,according to the [comment in QTBUG-109003](https://bugreports.qt.io/browse/QTBUG-109003?focusedCommentId=692790&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-692790), concerning the `qt_internal_add_qml_module` issue, maybe the problem is created by this line https://github.com/macports/macports-ports/blob/master/aqua/qt6/Portfile#L1324 `reinplace -q "s|\\\${CMAKE_CURRENT_LIST_DIR}|${qt_cmake_module_dir}/${cmakedir}|g" ${cmakefile}`
> 
> The macports' tbz2 package of qtbase (eg. https://packages.macports.org/qt6-qtbase/qt6-qtbase-6.2.4_0+openssl.darwin_20.x86_64.tbz2) shows that the original content which is:
> 
> ```
>     if (EXISTS "${CMAKE_CURRENT_LIST_DIR}/${INSTALL_CMAKE_NAMESPACE}${target}Macros.cmake")
> ```
> 
> is replaced by:
> 
> ```
>     if (EXISTS "/opt/local/libexec/qt6/lib/cmake/Qt6/${INSTALL_CMAKE_NAMESPACE}${target}Macros.cmake")
> ```
> 
> Thus qt won't install `Qt6QmlMacros.cmake` & `Qt6QmlBuildInternals.cmake`
> 
> Is this code snippet actually needed in the Portfile? Why? @jhoyt4 maybe?

Maybe.  

That line was inherited from the qt5 Portfile which I based this off of - the purpose of the line it to put the QT6 cmake files into a QT6 specific location in /opt/libexec so that they do not conflict with the other QT installed (e.g. qt4, qt5).

From your output, it looks like it's trying to put everything in the proper place.

**Comment by jhoyt4** [comment]
Understood.  Unfortunately, macports does things a bit differently.

This is due to macports need to support a lot of different macOS versions, eventually latching a qt version to a specific  macOS version when qt drops support of that os (this is at least one reason why there are so many qt5x's in the ports tree).  Currently, there's only one qt6, so your proposed change should not break anything for now.

If it does in fact solve the problem (still waiting on concrete evidence that it does), it will need to be noted in the portfile that this will need to be addressed at a future date.

BTW - I'm not a macports member - so one of the devs will need to chime in on this.

@reneeotten @kencu @mascguy@chrstphrchvz - thoughts?

**Comment by jhoyt4** [comment]
digging through the CI logs
`2022-11-28T20:19:03.6194860Z DEBUG: checksum file: /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_aqua_qt6/qt6-qtdeclarative/work/destroot/opt/local/libexec/qt6/lib/cmake/Qt6Qml/Qt6QmlBuildInternals.cmake`

It appears that the cmake file does get staged into desroot - so there's a good chance that your proposed solution will work.

Interestingly, the files also get copied into the correct (for macports) place - so that reinplace may be obsolete.

**Comment by tenzap** [comment]
I guess that by removing the reinplace, I have to increment the revision of all modules and qtbase as well. Renee?

**Comment by tenzap** [comment]
@reneeotten , are you ok with this fix?

**Comment by reneeotten** [comment]
> @reneeotten , are you ok with this fix?

I don't know - haven't had time to test this and see if it indeed fixes the issue without introducing others. Perhaps I can find some time this weekend but I don't make any promises.

**Comment by reneeotten** [comment]
@tenzap you claimed this worked, but it clearly doesn't build on the buildbots... can you please check the [failures](https://build.macports.org/builders/ports-13_x86_64-watcher/builds/1010) and fix?

**Comment by tenzap** [comment]
I'm sorry but I don't have a real mac at disposal and the one which I can access at times runs 10.11.

This built on the CI so I'm surprised this doesn't work on the builders.

Do you think this could be because of some components that were not rebuilt with the newer version (eg taking 6.2.4 rev0 instead of 6.2.4 rev1?)? Eg would qtmultimedia rev1 have used qtdeclarative rev0 ? This could explain the issue.

**Comment by tenzap** [comment]
@reneeotten, see #17059

File: aqua/qt6/Portfile [hunk 1]
```diff
@@ -1313,17 +1313,6 @@ post-destroot {
...
-        # if cmake finds configuration files in ${mp_prefix}/lib/cmake, CMAKE_CURRENT_LIST_DIR expands to ${mp_prefix}/lib/cmake/xxx
-        # cmake configuration files actually installed in ${qt_cmake_module_dir}/xxx
-        set cmakepath ${destroot}${qt_libs_dir}/cmake/
-        foreach cmakefile [ rglob ${cmakepath} *.cmake ] {
-            if { ![file isdirectory ${cmakefile} ] } {
-                set cmakedir [file dirname ${cmakefile}]
-                set cmakedir [string replace $cmakedir 0 [string length $cmakepath]-1]
-                reinplace -q "s|\\\${CMAKE_CURRENT_LIST_DIR}|${qt_cmake_module_dir}/${cmakedir}|g" ${cmakefile}
-            }
-        }
```
Comment (TellowKrinkle) [inline aqua/qt6/Portfile:1326]: Getting the following failure trying to configure a dolphin build:
```
CMake Error at /opt/local/lib/cmake/Qt6CoreTools/Qt6CoreToolsTargets.cmake:154 (message):
  The imported target "Qt6::moc" references the file

     "/opt/local/bin/moc"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/opt/local/lib/cmake/Qt6CoreTools/Qt6CoreToolsTargets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /opt/local/lib/cmake/Qt6CoreTools/Qt6CoreToolsConfig.cmake:37 (include)
  /opt/local/lib/cmake/Qt6Core/Qt6CoreDependencies.cmake:76 (find_package)
  /opt/local/lib/cmake/Qt6Core/Qt6CoreConfig.cmake:40 (include)
  /opt/local/lib/cmake/Qt6/Qt6Config.cmake:209 (find_package)
  Source/Core/DolphinQt/CMakeLists.txt:22 (find_package)
```
This looks awfully similar to what this code claims to fix...
(Putting this code back fixes it)

**Comment by jhoyt4** [comment]
@reneeotten @tenzap - it looks like removing the cmake edits in [commit  59633a2](https://github.com/macports/macports-ports/commit/59633a233bf6fbe3157248ae66fd45e14a178ffc) has caused new issues.

File: aqua/qt6/Portfile (hunk 1 above)
Comment (tenzap) [reply aqua/qt6/Portfile:1326]: How do you configure? With qmake or qt-cmake ?

You may see welle.io portfile for an example that uses qt-cmake (welle.io-devel subport actually) towards qt6

**Comment by tenzap** [comment]
> @reneeotten @tenzap - it looks like removing the cmake edits in [commit 59633a2](https://github.com/macports/macports-ports/commit/59633a233bf6fbe3157248ae66fd45e14a178ffc) has caused new issues.

For now there is no qmake6 portgroup. If you configure with cmake be sure to include first cmake pg, then qt6 pg. See welle.io-devel subport for a projet that builds fine on qt6.

File: aqua/qt6/Portfile (hunk 1 above)
Comment (tenzap) [reply aqua/qt6/Portfile:1326]: Restoring these lines is not the way to go IMO.

Comment (tenzap) [reply aqua/qt6/Portfile:1326]: See also #17071

Comment (TellowKrinkle) [reply aqua/qt6/Portfile:1326]: With normal `cmake`, with the `CMakeLists.txt` searching for qt with `find_package(Qt6 COMPONENTS Core Gui Widgets)`

Comment (TellowKrinkle) [reply aqua/qt6/Portfile:1326]: Note: This is not a port.  This is just me attempting to use macports' qt6 to satisfy a dependency when building a project from source.  AKA option 3 here:
https://github.com/macports/macports-ports/blob/223bfc67382a039df400d0b9323e75af984c3ed8/aqua/qt6/Portfile#L624

IMO this should continue to work.  I've posted an alternative way of making this work on #17071, if you'd prefer that.

Comment (tenzap) [reply aqua/qt6/Portfile:1326]: As per qt6 documentation: 
https://doc.qt.io/qt-6/cmake-build-on-cmdline.html

**Pull request merged**